README.rst
Lex
===

API for recommending news using content-based algorithms.


Try it
------

http://lex.cloud.tsuru.io/


How it works
------------

Currently Lex is a mock.


Success example
...............

::

    $ curl -i -X POST "http://lex.cloud.globoi.com/recommendation" -d  '{"userId": 1234, "userProvider": 2, "excludeIds": [123,5235,123], "dateStart": "1997-07-16T19:20:30.45+01:00", "dateEnd": "1997-07-16T19:20:30.45+01:00", "product":  "mobile", "limit": 2}'

    HTTP/1.1 200 OK
    date: Wed, 19 Feb 2014 13:17:01 GMT
    access-control-allow-origin: *
    content-type: text/html; charset=UTF-8
    content-length: 148
    server: TornadoServer/3.2
    connection: keep-alive

    [{"score":0.999,"documentId":6116038798098712381},{"score":0.666,"documentId":6502978799245377158},{"score":0.333,"documentId":5412347350701788586}]


Failure example
...............

::

    $ curl -i -X POST "http://lex.cloud.globoi.com/recommendation" -d "{}"

    HTTP/1.1 400 Bad Request
    date: Wed, 19 Feb 2014 13:16:05 GMT
    access-control-allow-origin: *
    content-type: application/json; charset=UTF-8
    content-length: 144
    server: TornadoServer/3.2
    connection: keep-alive

    {"error": "Expected the following keys: ['dateEnd', 'dateStart', 'excludeIds', 'limit', 'product', 'userId', 'userProvider'], but received: []"}(venus3)tati@tuxy ~/semrec/venus (master)


Healthcheck
...........

::

    $ curl -i -X GET "http://lex.cloud.globoi.com/healthcheck"

    HTTP/1.1 200 OK
    date: Wed, 19 Feb 2014 21:17:25 GMT
    content-length: 7
    etag: "451860c9eea89417e8c5baf88501beb03c8199f1"
    content-type: text/html; charset=UTF-8
    server: TornadoServer/3.2
    connection: keep-alive

    Atchim!


Version
.......

::

    $ curl -i -X GET "http://lex.cloud.globoi.com/version"

    HTTP/1.1 200 OK
    date: Wed, 19 Feb 2014 21:17:25 GMT
    content-length: 7
    etag: "451860c9eea89417e8c5baf88501beb03c8199f1"
    content-type: text/html; charset=UTF-8
    server: TornadoServer/3.2
    connection: keep-alive

    0.0.1

How to run
----------

In order to install dependencies and run the project: ::

    make setup
    make run


Check if service is running: ::

    curl http://0.0.0.0:8888/healthcheck

Is supposed to return ::

    Atchim!


How to run the tests
--------------------

We try hard to keep coverage to 100%, using unit tests whenever possible: ::

    make tests


Contributing
------------

Check if the style of the code is as expected: ::

    make style

And don't forget to provide tests in your patch! ;)


RELEASES.rst
Lex
===

0.0.1 (XX/02/2014)
------------------

- Mock, for validating integration


tests/unit/test_main.py
import unittest

from lex.main import main


class MainTestCase(unittest.TestCase):

    def test_main(self):
        response = main()
        expected = "Lex is running"
        self.assertEqual(response, expected)


tests/unit/test_params.py
import unittest

from lex import params


class ParamsTestCase(unittest.TestCase):

    def test_has_exception(self):
        self.assertTrue(params.ParamsException)

    def test_has_obligatory_keys(self):
        self.assertTrue(params.OBLIGATORY_KEYS)
        expected = [
            'dateStart',
            'product',
            'userId',
            'excludeIds',
            'dateEnd',
            'limit',
            'userProvider'
        ]
        self.assertEqual(params.OBLIGATORY_KEYS, expected)

    def test_validate_params_raises_exception(self):
        empty_params = {}
        with self.assertRaises(params.ParamsException) as exception:
            params.validate(empty_params)
        expected = "Expected the following keys: ['dateEnd', 'dateStart', 'excludeIds', 'limit', 'product', 'userId', 'userProvider'], but received: []"
        self.assertEqual(str(exception.exception), expected)

    def test_validate_params_succeeds(self):
        valid_params = {
            "userId": 1234,
            "userProvider": 2,
            "excludeIds": [123, 5235, 123],
            "dateStart": "1997-07-16T19:20:30.45+01:00",
            "dateEnd": "1997-07-16T19:20:30.45+01:00",
            "product": "mobile",
            "limit": 2
        }
        self.assertTrue(params.validate(valid_params))


tests/unit/test_settings.py
import unittest

from lex import settings


class SettingsTestCase(unittest.TestCase):

    def test_local_values(self):
        self.assertEqual(settings.DEBUG, True)
        self.assertEqual(settings.PORT, 8888)
        self.assertEqual(settings.HTML_PATH, 'docs/build/html')
        self.assertEqual(settings.STATIC_PATH, "docs/build/html/_static")


src/lex/main.py
"""
Lex starter and main loop.
"""
from tornado.web import Application, StaticFileHandler, URLSpec
from tornado.ioloop import IOLoop
from tornado.httpserver import HTTPServer
from tornado.log import app_log
from tornado.options import define, options, parse_command_line

from lex import settings
from lex.handlers import HealthcheckHandler, NewsHandler, RootHandler, VersionHandler


# Options
define("debug", default=settings.DEBUG, help="Enable or disable debug", type=bool)
define("port", default=settings.PORT, help="Run app on the given port", type=int)

# Temporary routes for serving HTML while Tsuru doesn't support a better way
# to serve static files. Eg. Nginx or other.
define("template_path", default=settings.HTML_PATH, help="Path where HTML doc files are", type=str)
define("static_path", default=settings.STATIC_PATH, help="Path where doc static files are ", type=str)


def create_app():
    """
    Create Lex instance of tornado.web.Application.
    """
    routes = [
        URLSpec(r'/?', RootHandler),
        URLSpec(r'/healthcheck/?', HealthcheckHandler),
        URLSpec(r'/version/?', VersionHandler),
        URLSpec(r'/recommendation/?', NewsHandler),
        URLSpec(r'/docs/?(.*)', StaticFileHandler, {'path': options["template_path"], 'default_filename': 'index.html'}),
        URLSpec(r'/_static/?(.*)', StaticFileHandler, {'path': options["static_path"], 'default_filename': 'favicon.ico'})
    ]
    return Application(routes, **options.as_dict())


def main():  # pragma: no cover
    """
    Run lex main loop.
    """
    parse_command_line()
    app_log.info("Lex is running!")
    application = create_app()
    server = HTTPServer(application)
    server.listen(options['port'])
    io_loop = IOLoop.instance()
    io_loop.start()


if __name__ == '__main__':
    main()
else:
    application = create_app()


src/lex/params.py
"""
Stuff related to parameters passed by the API clients.
"""

OBLIGATORY_KEYS = [
    'dateStart',
    'product',
    'userId',
    'excludeIds',
    'dateEnd',
    'limit',
    'userProvider'
]


class ParamsException(Exception):
    """
    Params-related exception. Has attribute:
        status_code = 400
    """
    status_code = 400


def validate(params_dict):
    """
    Validate client params dictionary.
    If it is valid, return True.
    Else, raise lex.params.ParamsException.
    """
    received_keys = sorted(params_dict.keys())
    expected_keys = sorted(OBLIGATORY_KEYS)

    if received_keys != expected_keys:
        msg = "Expected the following keys: {0}, but received: {1}"
        msg = msg.format(expected_keys, received_keys)
        raise ParamsException(msg)
    return True


src/lex/utils/git.py
"""
Module for retrieving git information, e.g. current hash.
"""

import subprocess


GET_BRANCH = 'git rev-parse --abbrev-ref HEAD'
GET_TAG = 'git describe --exact-match --tags HEAD'
GET_COMMIT = 'git rev-parse --verify HEAD'


def run(cmd):
    """
    Run a command on terminal and return the first line response
    """
    process = subprocess.Popen(cmd, shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
    return process.stdout.readline().split('\n')[0]


def get_version_label():
    """
    Return current branch or tag, if any, otherwise return <DEFAULT_VERSION>
    """

    tag = run(GET_TAG)
    branch = run(GET_BRANCH)
    label = tag or branch
    if label == 'HEAD':
        label = 'unstaged'
    return label


def get_version_hash():
    """
    Return current git hash.
    """
    return run(GET_COMMIT)


def get_code_version():
    """
    Return <git branch or tag> | <git hash> of source code.
    """
    label = get_version_label()
    commit = get_version_hash()
    version = u"%s | %s" % (unicode(label), unicode(commit))
    return version


src/lex/handlers.py
"""
Handlers for lex's API resources.
"""
import sys
import traceback

import ujson
from tornado.web import RequestHandler
from tornado_cors import CorsMixin

import lex
from lex.params import validate


class BaseHandler(CorsMixin, RequestHandler):
    """
    Base Handler, which provides:
    - Error handling
    - CORS (Cross-Origin Resource Sharing)
    """

    CORS_ORIGIN = "*"
    CORS_HEADERS = "Content-Type, Authorization"

    def _handle_request_exception(self, exception):
        self.log_exception(*sys.exc_info())
        if hasattr(exception, "status_code"):
            status = exception.status_code
            msg = exception.message
            self.send_error(status, message=msg)
        else:
            status = 500
            exc_info = sys.exc_info()
            msg = "<br>".join(traceback.format_exception(*exc_info))
            self.send_error(status, exec_info=exc_info, message=msg)

    def write_error(self, status_code, **kwargs):
        self.prepare()
        error_json = {"error": kwargs.get("message")}
        self.finish(error_json)


class HealthcheckHandler(RequestHandler):
    """
    Return "Atchim!" if service is running.
    """

    def get(self):
        self.write("Atchim!")


class RootHandler(RequestHandler):
    """
    Return "Atchim!" if service is running.
    """

    def get(self):
        self.write("My name is Lex. Lex Luthor.")


class VersionHandler(RequestHandler):
    """
    Return "Atchim!" if service is running.
    """

    def get(self):
        self.write(lex.__version__)


class NewsHandler(BaseHandler):
    """
    Return a list of recommendations provided the user of interest.
    """

    def post(self):
        params = ujson.loads(self.request.body)
        validate(params)
        recommendation_list = [
            {
                "documentId": "6116038798098712381",
                "score": 0.999
            },
            {
                "documentId": "6502978799245377158",
                "score": 0.666
            },
            {
                "documentId": "5412347350701788586",
                "score": 0.333
            }
        ]
        self.write(ujson.dumps(recommendation_list))
